Asynchronously wait for the engine to be drained

arteam opened the pull request:
Get rid of `UnsafePlainActionFuture` in `drainForClose` and offer async versions of the `Engine#close` and `Engine#flushAndClose` methods. 

Unfortunately, `IndexShard` still have to use `UnsafePlainActionFuture` to synchronously close an engine on the generic pool, but at least it's isolated. Other direct and indirect blocking calls on `Engine#close` via try-with-resources blocks should only happen in tests.

See ES-8644

arteam committed:
[DRAFT] Asynchronously wait for the engine to be drained

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

DaveCTurner commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1992,70 +1989,54 @@ protected final Releasable acquireEnsureOpenRef() {
         if (isClosing.compareAndSet(false, true) == false) {
             logger.trace("drainForClose(): already closing");
-            return false;
+            return;
```
Hmm this seems like it leaks `listener` without completing it. That may be what you intend, but it's pretty trappy. Could we fail it with an `AlreadyClosedException` perhaps?

DaveCTurner commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1992,70 +1989,54 @@ protected final Releasable acquireEnsureOpenRef() {
     /**
      * Flush the engine (committing segments to disk and truncating the translog) and close it.
      */
     public void flushAndClose() throws IOException {
```
Are you sure it works to leave this method as synchronous? It's called from `org.elasticsearch.index.shard.IndexShard#close` and when it returns it will execute the subsequent `finally` block, which may now be happening too soon while it's still draining.

arteam replied:
Yes, that's a very good point. It seems there's indeed an expectation that `Engine#flushAndClose` and `Engine#close` are synchronous. I wonder if we need to make them asynchronous too, because otherwise we would still need to use `PlainActionFuture`

arteam committed:
Fail the listener is it's already closing

arteam committed:
Async Engine#close and Engine#flushAndClose

arteam committed:
Merge remote-tracking branch 'origin/main' into async-drain-for-close

arteam committed:
Run spotless

DaveCTurner commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1992,70 +1989,66 @@ protected final Releasable acquireEnsureOpenRef() {
 
     @Override
     public void close() throws IOException {
+        close(ActionListener.noop());
```
I don't think we can make this a fire-and-forget action. At best, it's very trappy that this method would return before the engine is closed.

Do we really need `Engine implements Closeable` anyway? If we dropped that we could make `close()` properly async.

arteam replied:
> Do we really need Engine implements Closeable anyway? If we dropped that we could make close() properly async.

Thank you! I was stuck on this point, too. Let me see whether it would be possible to get rid of the sync `flushAndClose` and `close` methods.

arteam commented:
It seems to be quite hard to make the `close` and `flushAndClose` methods asynchronous. `Engine` extends `Closeable` and is closed indirectly in this way in a lot of places. `flushAndClose` callers also expect it to be synchronous and the engine to be closed after they are finished, e.g `IndexShard#close` that wraps the `flushAndClose` call with a try-finally block where it closes internal references.

Despite that `drainForClose` is now asynchronous, both these methods preserve synchronous invariant due to waiting on the `awaitPendingClose` latch. Still, I really don't like that we pass `ActionListener.noop()` as the close listener. It seems to be that we have to pass some variant of `Future` to make explicit that we have to wait until the close listener in `drainForClose` finishes.

DaveCTurner commented:
> `Engine` extends `Closeable` and is closed indirectly in this way in a lot of places.

In production code, or just in tests? In tests we can introduce a blocking wrapper around an async method using a `Future`.

> e.g `IndexShard#close` that wraps the `flushAndClose` call with a try-finally block where it closes internal references.

Yeah but that's all happening on a background executor, we can move the `finally` block into a `runBefore` around the outer listener I think.

arteam committed:
Async flushAndClose

arteam committed:
Merge remote-tracking branch 'origin/main' into async-drain-for-close

arteam committed:
Always complete flushAndClose listener

Always complete close latch

Run spotless

Always complete flushAndClose listeners

Use public Engine#close(listener)

arteam committed:
Wait until close(listener) completes

arteam committed:
Revert "Wait until close(listener) completes"

This reverts commit afb4723045e974b358e3794c1304ae680cd93325.

arteam committed:
Don't use blocking PlainActionFuture

arteam committed:
Override the listener call on onClose

arteam committed:
Use UnsafeExecutor instead of PlainExecutor to avoid deadlocks

arteam committed:
Do we still run on the cluster applier thread?

arteam committed:
Figure out where close on Engine is called from

arteam committed:
Check that Engine#close is only called from tests

arteam committed:
Improve comments

arteam committed:
Merge remote-tracking branch 'origin/main' into async-drain-for-close

arteam committed:
Use RefCountingListener

arteam committed:
Add more comments

arteam committed:
Merge remote-tracking branch 'origin/main' into async-drain-for-close

arteam commented:
@DaveCTurner Thank you! I've tried to make sure that `Engine#close` only gets called in tests and migrated the rest of the code to async versions of `close` and `flushAndClose`. There's still a few blocking calls left in `IndexShard` where it seems to quite hard to the async shard closing.

DaveCTurner commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1992,72 +1990,79 @@ protected final Releasable acquireEnsureOpenRef() {
 
     @Override
     public void close() throws IOException {
+        // TODO Called indirectly via try-with-resources or `IOUtils.close` from tests, Ideally, we would remove this method,
+        // and replace it with a blocking `EngineTestCase.close()` call, but it's used in a ton of tests and require
+        // refactoring, so we just check that the method is called from a test thread
```
I'd rather we did this work up-front first to remove the `Closeable` interface from `Engine`. It may be rather noisy but it shouldn't be particularly complicated.

arteam replied:
Yeah, I tried to do that, there were a lot of changes required, especially in `InternalEngineTests` which is 7800 lines of code. Let me spin up a separate PR in trying to remove `Closeable` from `Engine`, maybe I will have some success with automatic refactoring tools :+1:

arteam replied:
I've submitted #111094

DaveCTurner commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -4368,4 +4386,16 @@ public void waitForPrimaryTermAndGeneration(long primaryTerm, long segmentGenera
+    private static void closeEngine(Engine engine) throws IOException {
+        if (engine != null) {
+            // TODO Ideally we would convert `closeEngine` calls to async `close` calls and remove UnsafePlainActionFuture
+            final var future = new UnsafePlainActionFuture<Void>(ThreadPool.Names.GENERIC);
```
Introducing a new `UnsafePlainActionFuture` here doesn't really improve things much. Could we instead first make all the engine-closing calls take a listener (but in practice remain synchronous) with one or more preliminary PRs, and then we can flip it to an async implementation without fuss.

arteam replied:
I hated to use it here, too. Indeed, let me try to get rid of all synchronous calls in a separate PR :+1:

arteam commented:
The PR got split into #111094, #111116, #111582 and #111796
